### ISSUE TITLE
[core] Task concurrency limited when pull manager has requests queued

### ISSUE THREAD

**Issue by stephanie-wang** [ray-project/ray#17628]
### What is the problem?

*Ray version and other system information (Python version, TensorFlow version, OS):* 2.0dev

The pull manager will only pull task args up to the available capacity in the local object store. However, sometimes space can also be made by spilling local objects. So a node  with lots of local objects can execute much fewer tasks than expected concurrently, according to its available CPU and object store memory.

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

On my laptop, the following script runs in ~3s. Uncommenting the line below, it runs in 1s because the remote node spills all of its local objects and can run the second round of tasks in parallel.

```python
@pytest.mark.parametrize(
    "ray_start_cluster_head", [{
        "num_cpus": 0,
        "object_store_memory": 75 * 1024 * 1024,
    }],
    indirect=True)
def test_maximize_concurrent_pulls(ray_start_cluster_head):
    cluster = ray_start_cluster_head
    cluster.add_node(num_cpus=8, object_store_memory=75 * 1024 * 1024)

    @ray.remote
    def put():
        return np.random.rand(1 * 1024 * 1024)  # 8 MB data

    @ray.remote
    def f(x):
        time.sleep(1)
        return

    remote_refs = [put.remote() for _ in range(7)]
    #ray.get([f.remote(x) for x in remote_refs])  # Force objects on the remote node to spill, making room for task args.
    local_refs = [ray.put(1 * 1024 * 1024) for _ in range(8)]
    remote_tasks = [f.remote(x) for x in local_refs]

    start = time.time()
    ray.get(remote_tasks)
    end = time.time()
    print(end - start)
```

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

**Comment by stephanie-wang** [comment]
@rkooo567 @scv119, this is a performance issue not a stability one, but I think it could lead to big slowdowns when there are lots of cores that go unutilized.

**Comment by rkooo567** [comment]
I have one question! Does it happen only when we pull for the first time in a while? For example, if we are actively pulling the spilling is already started, so we don't see this issue?

**Comment by stephanie-wang** [comment]
I don't think it matters when the last time we pulled was. The problem is that the PullManager is trying to limit how many active pulls there are so that it doesn't trigger OOM, but by doing that, it also doesn't ever trigger spilling. So we might end up pulling much less memory than we could actually use.

**Comment by ericl** [comment]
I don't think this is right, we should be spilling continuously at max throughput as long as there is any CreateRequestQueue pressure. This was the main point of the refactorings I merged earlier.

So in this scenario, we are already spilling as fast as we can. Is this not the case?

**Comment by stephanie-wang** [comment]
Currently we are only spilling if the CreateRequestQueue exceeds its quota, which will trigger OOM handling in the object store. But queued task args respect the quota, so we never trigger spilling.

Discussed offline, the fix is to lower the spill threshold so that we spill before the object store is 100% full. This should work as long as we set the threshold low enough.

**Comment by ericl** [comment]
It sounds good, but it's not right we don't trigger spilling on queued task
args. We always fetch at least one bundle so that will trigger spilling.
The task itself might not get scheduled but the bundle will be fetched.

**Comment by stephanie-wang** [comment]
That will only trigger spilling if the one bundle puts us over the memory quota.

**Comment by rkooo567** [comment]
Okay. So I've worked on this a bit. 

- Script here has some issues. I modified it.
```python3
@pytest.mark.parametrize(
    "ray_start_cluster_head", [{
        "num_cpus": 0,
        "object_store_memory": 75 * 1024 * 1024,
        "_system_config": {
            "worker_lease_timeout_milliseconds": 0
        }
    }],
    indirect=True)
def test_maximize_concurrent_pulls(ray_start_cluster_head):
    cluster = ray_start_cluster_head
    cluster.add_node(num_cpus=8, object_store_memory=75 * 1024 * 1024)

    @ray.remote
    class RemoteObjectCreator:
        def put(self, i):
            return np.random.rand(i * 1024 * 1024)  # 8 MB data

        def idle(self):
            pass

    @ray.remote
    def f(x):
        print(x)
        time.sleep(1)
        return

    remote_obj_creator = RemoteObjectCreator.remote()
    remote_refs = [remote_obj_creator.put.remote(1) for _ in range(7)]
    # Make sure all objects are created.
    ray.get(remote_obj_creator.idle.remote())
    # Trigger spilling
    remote_refs.append(remote_obj_creator.put.remote(1))

    local_refs = [ray.put(np.random.rand(1 * 1024 * 1024)) for _ in range(8)]
    print(local_refs)
    remote_tasks = [f.remote(x) for x in local_refs]

    start = time.time()
    ray.get(remote_tasks)
    end = time.time()
    print(end - start)
```
There are 2 discoveries. 

1. After I modified it, I actually couldn't reproduce the issue in the small scale because eventually the spilling is triggered because there's always 1 bundle that's pulled (so eventually spilling is triggered at some point).
  - Without faster spilling
    - Pull upto the max quota -> 1 bundle pulling -> 1 bundle pulling triggers the spilling -> many tasks
  - With faster spilling
    - Pull upto the max quota -> spilling (pipelined)-> pull upto max quota. 

I think lowering the threshold is definitely good for performance when the memory usage is large, so if there's no objection, I'd just proceed with that.

**Comment by rkooo567** [comment]
This is another scenario I found.

This happens 100% of time when pipeline spilling wasn't enabled, and 50% of time when it is enabled (we should eventually solve it by better retry protocol). TL;DR this block is triggered quite often which made it wait 10 additional seconds. 
https://github.com/ray-project/ray/blob/5571d2436da42438c26967a5d1211a2ebbeffe42/src/ray/object_manager/object_buffer_pool.cc#L97

Object store mem: 64MB/70MB
Object: 8MB with 2 chunks

- chunk index 1 receives at thread A
- buffer pool attempts to spill & create an object 
- chunk index 0 receives at thread B
- buffer pool attempts to spill & create an object 
- Object is created by chunk index 0 (since it is FIFO)
- But chunk index 1 gets reply from the plasma store first.
- chunk index 0 failed at https://github.com/ray-project/ray/blob/5571d2436da42438c26967a5d1211a2ebbeffe42/src/ray/object_manager/object_buffer_pool.cc#L97
- chunk index 1 creates the object successfully.
- Since chunk index 0 failed, it should wait 10 more seconds to retry.

This can be solved by either 1. better retry protocol we are planning to do. 2. cache the request until object is created (this is not ideal because it can cause large memory usage) and seal it as soon as it is created.

**Comment by stephanie-wang** [comment]
Hmm I'm not sure why spilling still gets triggered when pulling the 1 bundle would still keep the pull manager under quota. Maybe it's just because the estimation of memory remaining is not exact?

@rkooo567 can you file a separate issue for the object store race? It looks like we could fix it just by not releasing the lock during the Create call but we should track it separately.

**Comment by rkooo567** [comment]
Let me create an issue for the second observation. One of concerns of holding a lock is that the object manager server receives are now blocked by spilling (which is probably fine if pipeline is on?). We can discuss more in a separate issue.

**Comment by stephanie-wang** [comment]
I looked into the script a bit more and it looks like this is actually not an issue after the first task gets dispatched. At that point, we're back to pulling no args, so then we'll trigger spilling. The sequence is something like this:
1. pull args for task 1, under quota so we don't spill or pull args for task 2 even though we have available CPU
2. pin args for task 1, dispatch task 1
3. now we have no task args being pulled, so we pull args for task 2. this puts us over quota and we spill

It's probably still good to lower the spilling threshold by default, but I'm going to close this issue.
